### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by nvgrw**
A pull request to allow project mentors to review the code

### CODE REVIEW

File: compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AMD64Assembler.java [hunk 1]
```diff
@@ -949,10 +949,10 @@ public boolean check(AMD64 arch, int l, Register r, Register v, Register m, Regi
                     break;
             }
             if (r != null) {
-                assert r.getRegisterCategory().equals(rCategory);
+                assert r.getRegisterCategory().equals(rCategory) : "r register " + r.name + " with category " + r.getRegisterCategory() + " does not match expected category " + rCategory;
```
Comment (rramakrishna) [inline compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AMD64Assembler.java]: Test, please ignore.

File: compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java [hunk 2]
```diff
@@ -74,18 +83,29 @@ private static AMD64Kind twice(AMD64Kind kind) {
             case SINGLE:
                 return AMD64Kind.QWORD;
             default:
-                return kind;
+                throw GraalError.shouldNotReachHere("Unable to double AMD64Kind '" + kind.toString() + "'");
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:86]: A personal taste thing but I prefer format strings over string concatenation

File: compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java [hunk 3]
```diff
@@ -101,26 +121,62 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
...
                     masm.movdqu(asRegister(result), address);
                     break;
-                case 32: // TODO: Avoid using a constant here.
+                case YMM:
                     assert ((AMD64) masm.target.arch).getFeatures().contains(AMD64.CPUFeature.AVX) : "AVX is unsupported";
                     masm.vmovdqu(asRegister(result), address);
                     break;
                 default:
-                    throw GraalError.shouldNotReachHere();
+                    throw GraalError.shouldNotReachHere("Unsupported constant size '" + AVXKind.AVXSize.fromBytes(byteBuffer.capacity()));
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java]: String concat

File: compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java [hunk 4]
```diff
@@ -101,26 +121,62 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
...
+     * @param src The value we want to move from. Can be a register, stack variable, or constant.
+     * @param srcKind The AMD64Kind of the src value.
+     * @param scratch A scratch register to use if necessary. Must be allocated by the caller as an @Temp.
+     */
+    private static void move(CompilationResultBuilder crb, AMD64MacroAssembler masm, AMD64Address dst, Value src, AMD64Kind srcKind, Register scratch) {
+        if (isRegister(src)) {
+            reg2addr(crb, masm, srcKind, dst, asRegister(src));
+        } else if (isStackSlot(src)) {
+            // We move first to the scratch register, then to our destination.
+            // Is there a faster solution?
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java]: If this is something that you wish to revisit, please mark it TODO

File: compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java [hunk 5]
```diff
@@ -101,26 +121,62 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
...
+     * from a memory location is in StoreStackOp, and the target in that case is always an array.
+     * If we want to implement a stack-based unpacking instruction, though, this might be worth
+     * considering.
+     *
+     * @param crb
+     * @param masm
+     * @param dst The address we want to move to.
+     * @param src The value we want to move from. Can be a register, stack variable, or constant.
+     * @param srcKind The AMD64Kind of the src value.
+     * @param scratch A scratch register to use if necessary. Must be allocated by the caller as an @Temp.
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java]: nit: a `@Temp`

File: compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java [hunk 6]
```diff
@@ -190,27 +246,32 @@ private static void const2addr(CompilationResultBuilder crb, AMD64MacroAssembler
...
+     */
     public static final class LoadStackOp extends AMD64LIRInstruction {
-
-        private static final int YMM_LENGTH_IN_BYTES = 32;
         public static final LIRInstructionClass<LoadStackOp> TYPE = LIRInstructionClass.create(LoadStackOp.class);
 
+        // Our destination should always be a register.
         @Def({REG}) private AllocatableValue result;
+
+        // We need a scratch register for any possible memory-to-memory moves (see move(...) above).
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java]: I think there's some way to link to other methods? Might be worth investigating the exact Javadoc syntax for this but they use it a bunch throughout the codebase

File: compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java [hunk 7]
```diff
@@ -220,68 +281,98 @@ protected LoadStackOp(LIRInstructionClass<? extends LoadStackOp> c, LIRGenerator
...
+            // If we can fill an entire YMM register, we can just perform a single move and skip
+            // more complicated packing logic.
+            // TODO: support XMM-only platforms
+            if (sizeInBytes == YMM.getBytes()) {
                 masm.vmovdqu(asRegister(result), inputAddress);
             } else {
-                // Lowest multiple of YMM_LENGTH_IN_BYTES that can fit all elements.
-                int amt = (int) Math.ceil(((double) sizeInBytes) / YMM_LENGTH_IN_BYTES) * YMM_LENGTH_IN_BYTES;
+                // Allocate a register-width of stack space for us to dump our values to.
+                masm.subq(rsp, YMM.getBytes());
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java]: Why restrict to YMM?

File: compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java [hunk 8]
```diff
@@ -220,68 +281,98 @@ protected LoadStackOp(LIRInstructionClass<? extends LoadStackOp> c, LIRGenerator
...
+    /**
+     * This operation stores a vector register into an array. It does so with
+     * the aid of some temporary stack space to minimize expensive vector instructions.
+     */
     public static final class StoreStackOp extends AMD64LIRInstruction {
-
-        private static final int YMM_LENGTH_IN_BYTES = 32;
         public static final LIRInstructionClass<StoreStackOp> TYPE = LIRInstructionClass.create(StoreStackOp.class);
 
+        // The array we're writing to.
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java:384]: Not necessarily an array?

File: compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java [hunk 9]
```diff
@@ -303,57 +397,82 @@ public void emitCode(CompilationResultBuilder crb, AMD64MacroAssembler masm) {
...
+     */
     public static final class PackStackOp extends AMD64LIRInstruction {
 
         private static final int YMM_LENGTH_IN_BYTES = 32;
         public static final LIRInstructionClass<PackStackOp> TYPE = LIRInstructionClass.create(PackStackOp.class);
 
+        // Our destination should always be a register.
         @Def({REG}) private AllocatableValue result;
+
+        // We need a scratch register for any possible memory-to-memory moves (see move(...) above).
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.lir.amd64/src/org/graalvm/compiler/lir/amd64/vector/AMD64Packing.java]: Linking thing again

File: compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AVXKind.java [hunk 10]
```diff
@@ -41,26 +41,36 @@
...
-                    return 8;
-                case XMM:
-                    return 16;
-                case YMM:
-                    return 32;
-                case ZMM:
-                    return 64;
+        DWORD(4),
+        QWORD(8),
+        XMM(16),
```
Comment (nvgrw) [inline compiler/src/org.graalvm.compiler.asm.amd64/src/org/graalvm/compiler/asm/amd64/AVXKind.java:46]: Are we sure that this isn't somewhere else in JVMCI/Graal? There are constants used for the AMD64 registers in various places (`xmm0`, etc) - is it worth repeating this?

**Review by nvgrw — COMMENTED** [review]
just some comments to perhaps consider :D

**Pull request closed without merging**